presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoSession.java

```diff
@@ -0,0 +1,572 @@
+            iterable.batchSize(cursorBatchSize);
+        }
+
+        return iterable.iterator();
+    }
+
+    @VisibleForTesting
+    static Document buildQuery(TupleDomain<ColumnHandle> tupleDomain)
+    {
+        Document query = new Document();
+        for (Map.Entry<ColumnHandle, Domain> entry : tupleDomain.getDomains().entrySet()) {
+            MongoColumnHandle column = checkType(entry.getKey(), MongoColumnHandle.class, "columnHandle");
+            query.putAll(buildPredicate(column, entry.getValue()));
+        }
+
+        return query;
+    }
+
+    private static Document buildPredicate(MongoColumnHandle column, Domain domain)
+    {
+        String name = column.getName();
+        if (domain.getRanges().isNone() && domain.isNullAllowed()) {
+            return documentOf(name, isNullPredicate());
+        }
+        if (domain.getRanges().isAll() && !domain.isNullAllowed()) {
+            return documentOf(name, isNotNullPredicate());
+        }
+
+        List<Object> singleValues = new ArrayList<>();
+        List<Document> disjuncts = new ArrayList<>();
+        for (Range range : domain.getRanges()) {
+            if (range.isSingleValue()) {
+                singleValues.add(range.getSingleValue());
+            }
+            else {
+                Document rangeConjuncts = new Document();
+                if (!range.getLow().isLowerUnbounded()) {
+                    switch (range.getLow().getBound()) {
+                    case ABOVE:
+                        rangeConjuncts.put(GT_OP, range.getLow().getValue());
+                        break;
+                    case EXACTLY:
+                        rangeConjuncts.put(GTE_OP, range.getLow().getValue());
+                        break;
+                    case BELOW:
+                        throw new IllegalArgumentException("Low Marker should never use BELOW bound: " + range);
+                    default:
+                        throw new AssertionError("Unhandled bound: " + range.getLow().getBound());
+                    }
+                }
+                if (!range.getHigh().isUpperUnbounded()) {
+                    switch (range.getHigh().getBound()) {
+                    case ABOVE:
+                        throw new IllegalArgumentException("High Marker should never use ABOVE bound: " + range);
+                    case EXACTLY:
+                        rangeConjuncts.put(LTE_OP, range.getHigh().getValue());
+                        break;
+                    case BELOW:
+                        rangeConjuncts.put(LT_OP, range.getHigh().getValue());
+                        break;
+                    default:
+                        throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                    }
+                }
+                // If rangeConjuncts is null, then the range was ALL, which should already have been checked for
+                checkState(!rangeConjuncts.isEmpty());
```

Comment:
Guava recently added a `verify()` (in `Verify`) for these types of runtime assertions.